### PR TITLE
changes to py-torch recipe to enable rocm build

### PR DESCRIPTION
**Pull request by kolamsrinivas**
@adamjstewart , Made changes as per your review comments -
To your questions => Will other versions of PyTorch build with ROCM, or do we need to backport this patch to older releases to get that working?
When i did a spack install py-torch +rocm -cuda -cudnn -nccl  , the build failed. On looking further i found there was a commit that recently went to fix. Hence i took the fix and applied the patch.

I also tested the build for spack install py-torch@master +cuda +cudnn +nccl cuda_arch=70 ; spack install py-torch +cuda +cudnn +nccl cuda_arch=70.

### CODE REVIEW

**Review by adamjstewart — APPROVED** [review]

**Comment by adamjstewart** [comment]
Flake8:
```
var/spack/repos/builtin/packages/py-torch/package.py:9: [E302] expected 2 blank lines, found 1
var/spack/repos/builtin/packages/py-torch/package.py:176: [W293] blank line contains whitespace
var/spack/repos/builtin/packages/py-torch/package.py:251: [W293] blank line contains whitespace
var/spack/repos/builtin/packages/py-torch/package.py:255: [W293] blank line contains whitespace
```

File: var/spack/repos/builtin/packages/py-torch/package.py [hunk 1]
```diff
@@ -68,6 +68,7 @@ class PyTorch(PythonPackage, CudaPackage):
 
     variant('cuda', default=True, description='Build with CUDA')
     variant('cudnn', default=True, description='Enables the cuDNN build')
+    variant('rocm', default=False, description='Build with ROCm build')
```
Comment (becker33) [inline var/spack/repos/builtin/packages/py-torch/package.py:72]: The way this is currently implemented, `py-torch +rocm` still can depend on CUDA. Is that a reasonable way for this to be implemented?

File: var/spack/repos/builtin/packages/py-torch/package.py [hunk 2]
```diff
@@ -297,6 +305,11 @@ def enable_or_disable(variant, keyword='USE', var=None, newer=False):
         enable_or_disable('zstd', newer=True)
         enable_or_disable('tbb', newer=True)
 
+    @run_before('install')
+    def build_amd(self):
+        if '+rocm' in self.spec:
+            python(os.path.join('tools', 'amd_build', 'build_amd.py'))
```
Comment (becker33) [inline var/spack/repos/builtin/packages/py-torch/package.py:311]: This should probably be part of the build process not a pre-install step

**Review by becker33 — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/py-torch/package.py (hunk 1 above)
Comment (becker33) [reply var/spack/repos/builtin/packages/py-torch/package.py:72]: The alternative is to add a conflict so `py-torch +rocm +cuda` is invalid.

File: var/spack/repos/builtin/packages/py-torch/package.py (hunk 2 above)
Comment (adamjstewart) [reply var/spack/repos/builtin/packages/py-torch/package.py:311]: PyTorch is a bit of an odd case here. Normally, `PythonPackage` builds have 2 stages: build and install. For PyTorch (and many other Python libraries), when you run `python setup.py build` and `python setup.py install`, the second command will redo all of the work completed in the first stage, so I removed the build phase from this package. This drastically improves installation times.

File: var/spack/repos/builtin/packages/py-torch/package.py (hunk 1 above)
Comment (kolamsrinivas) [reply var/spack/repos/builtin/packages/py-torch/package.py:72]: Added the conflicts('+rocm', when '+cuda')

**Comment by kolamsrinivas** [comment]
adamjstewart , i have made changes as per the review comments and addressed the flake8 issues that were reported. Can you please let me know if this can be merged .

**Comment by adamjstewart** [comment]
LGTM, restarted some failing tests, remind me to merge when all the tests finish.

**Comment by kolamsrinivas** [comment]
adamjstewart , i see the tests are completed.. in the last cycle, the flake8 were passing. Let me know if i need to do anything here

**Comment by adamjstewart** [comment]
Not sure why flake8 is failing again, reran the tests. Also need @becker33 to finish re-reviewing the PR.

**Comment by kolamsrinivas** [comment]
adamjstewart   I see from the log for flake8 tests - var/spack/repos/builtin/packages/smartdenovo/package.py:21: [E128] continuation line under-indented for visual indent.  think it is not related to this change.

**Comment by kolamsrinivas** [comment]
becker33, can you please review and approve the changes.

**Comment by adamjstewart** [comment]
@alalazo I believe there is a bug in our `flake8` tests. I'm seeing:
```
flake8: running flake8 code checks on spack.

Modified files:
  lib/spack/spack/build_systems/cuda.py
  var/spack/repos/builtin/packages/cuda/package.py
  var/spack/repos/builtin/packages/py-torch/package.py
  var/spack/repos/builtin/packages/smartdenovo/package.py
=======================================================
var/spack/repos/builtin/packages/smartdenovo/package.py:21: [E128] continuation line under-indented for visual indent
```
but this PR didn't modify `smartdenovo` or `cuda`, those were other recently merged PRs. It seems to be pulling the wrong base branch or something.

**Review by becker33 — APPROVED** [review]

**Pull request merged**